src/lib.rs

```diff
@@ -81,190 +84,154 @@ impl AgentState {
                 }
-            }
-            ast::Command::JumpTrue(next, expr) => {
-                let prim = self.evaluate_mut(expr)?;
-                match prim {
-                    pi @ Primitive::Integer(_) => {
-                        Err(format!("condition is non-boolean: {:?}", &pi))
-                    }
-                    Primitive::Boolean(false) => todo!(),
-                    Primitive::Boolean(true) => {
-                        self.pc = next;
-                        Ok(vec![])
-                    }
+                other => return Err(format!("invalid type [{:?}] for x", other)),
+            },
+            "y" => match value {
+                Primitive::Integer(i) if i > 0 => {
+                    self.coords = Coordinates(self.coords.x(), i as u32)
                 }
+                other => return Err(format!("invalid type [{:?}] for y", other)),
+            },
+            "color" => match value {
+                Primitive::Integer(i) if i > 0 => self.color = i as u32,
+                other => return Err(format!("invalid type [{:?}] for color", other)),
+            },
+            _ => {
+                self.vars.insert(id, value);
+            }
+        };
+        self.pc += 1;
+        Ok(vec![])
+    }
+}
+
+impl EvaluateMut<ast::FaceCmd> for AgentState {
+    type Output = Result<Vec<Coordinates>, String>;
+
+    fn evaluate_mut(&mut self, operation: ast::FaceCmd) -> Self::Output {
+        let ast::FaceCmd(new_direction) = operation;
+
+        self.direction = new_direction;
+        self.pc += 1;
+        Ok(vec![])
+    }
+}
+
+impl EvaluateMut<ast::TurnCmd> for AgentState {
+    type Output = Result<Vec<Coordinates>, String>;
+
+    fn evaluate_mut(&mut self, operation: ast::TurnCmd) -> Self::Output {
+        let ast::TurnCmd(rotations) = operation;
+
+        let original_direction = self.direction as i32;
+        self.direction = ast::Direction::from(original_direction + rotations);
+        self.pc += 1;
+        Ok(vec![])
+    }
+}
+
+impl EvaluateMut<ast::MoveCmd> for AgentState {
+    type Output = Result<Vec<Coordinates>, String>;
+
+    fn evaluate_mut(&mut self, operation: ast::MoveCmd) -> Self::Output {
+        let ast::MoveCmd(steps) = operation;
+        let orientation = self.direction;
+        let origin = self.coords;
+
+        let touched_cells: Vec<Coordinates> = (0..=steps)
+            .into_iter()
+            .map(|offset| move_in_direction(offset, orientation, origin))
+            .collect();
+        let end = touched_cells.last().copied().unwrap_or(origin);
+
+        self.coords = Coordinates(end.x(), end.y());
+        self.pc += 1;
+        Ok(touched_cells)
+    }
+}
+
+/// Updates coordinates to represent a move of n steps in a given direction
+fn move_in_direction(steps: u32, direction: ast::Direction, origin: Coordinates) -> Coordinates {
+    match direction {
+        ast::Direction::N => Coordinates(origin.x(), origin.y() + steps),
```

Comment:
North should move negatively across Y, and South should move positively across Y (applies to diagonal directions as well)

Reply:
Inverted N and S